Camera monitor for Modular Computers - Episode II

atlantiscze opened the pull request:
Hopefully fixed version of #13210 - see this PR for information. This is same from content point of view, just (hopefully!) without the map corruption problems.

atlantiscze committed:
Initial implementation of Camera Monitor program

atlantiscze committed:
Finishes work on camera monitor program
- Completely removes old camera monitor computer from the code and replaces it with modular computer variant.
- Camera monitor is now part of most station consoles. Access is network-based, so networks you can view depend on your ID.
- (Minor tweak) NTNet DOS program now uses a define multiplier instead of hardcoded number. Also reduced this multiplier to 5 (from 10) due to recent major buffs to ntnet speeds. The rate at which it could crash a relay was way too big.
- A special variant of camera monitor program was added, available to antagonists. It has access to all station networks without needing ID, and also to mercenary, ERT and crescent camera networks. Has a tiny chance to trigger an IDS alarm when IDS is active during prolonged operation. This variant is MUCH larger in terms of file size, making it hard to use on tablets.
- Another special variant of this program exists for ERT. It is slightly larger in terms of file size, and allows usage of ERT and crescent networks, if you have ERT access on your ID.
- Adds few extra types for mapped-in consoles - ERT console that contains essential emergency programs, and Mercenary console that contains hacked camera monitor and comes pre-emagged so you can download any antag programs you want.
- Changelog entry
- WILL CAUSE COMPILE ERRORS due to map changes. Next commit will fix these.

atlantiscze committed:
Replaces invalid typepaths in the map

atlantiscze committed:
Re-adds few consoles where telescreens were
- Most notably, interrogation and toxins testing. Also replaces few old alarm consoles with modular variants that can run the alarm monitor program (which has no access restriction)

atlantiscze committed:
Ports Camera monitor commits, without the map changes that caused corruption

Raptor1628 commented:
Now we just need records.

atlantiscze commented:
This should be in working condition, however, if a merge conflict shows up on torch-10 again it will take few hours for me to fix as the phython marpmerger doesn't work for me either.. So please, if possible, merge before that occurs.

atlantiscze committed:
Fixes Exodus map errors.

mwerezak commented:
...are you using Python 3?

atlantiscze commented:
Using python 3.5.1

mwerezak commented:
I don't know then, I haven't encountered any issues like that with the python mapmerger so far.

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-06-22-FixedCameras

atlantiscze commented:
Has there been any progress resolving the Torch map corruption? I recall something was said back on the staff meeting, that there should be a map fixing script around somewhere. Unfortunately for me it's a linux-only thing so i still can't do anything about the problem.

Kasuobes commented:
What exception is mapmergepy throwing at you?

atlantiscze commented:
This is translated from my local system language:

python.exe - application error
Application (0xc000007b) was unable to be started correctly. Click OK to close the application

A generic windows error it seems.

Kasuobes commented:
Try uninstalling/reinstalling python 3.5.2? Google says it might be a 32-bit vs 64-bit problem: http://stackoverflow.com/questions/20650596/cannot-open-python-error-0xc000007b

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-06-22-FixedCameras

- This may or may not work. Hopefully it will.

atlantiscze committed:
Alternative workaround
- Given this PR has been held up for about three months now, and as nobody has found a way to fix the regular mapmerge's problems, i've decided to use an alternative workaround.
- Types with matching paths as the old consoles/circuits/etc. that exist on the map have been added. Some immediately qdel() themselves, some spawn a relevant modular console on their location. A crude solution, but it works. Exodus map does not use any of these, but Torch has bunch of them, as i am unable to change Torch Z levels without running into the problem again.

atlantiscze commented:
- Given this PR has been held up for about three months now, and as nobody has found a way to fix the regular mapmerge's problems, i've decided to use an alternative workaround.
- Types with matching paths as the old consoles/circuits/etc. that exist on the map have been added. Some immediately qdel() themselves, some spawn a relevant modular console on their location. A crude solution, but it works. Exodus map does not use any of these, but Torch has bunch of them, as i am unable to change Torch Z levels without running into the problem again.

Raptor1628 commented:
Have you tried using the python mapmerger on the torch? That seemed to fix things right up for me. If this is the only working way, let's get this show on the road.

atlantiscze commented:
Despite several attempts, i did not manage to get the python mapmerge to work on my computer. I am not sure what is the problem. Either way, this solution works to the end user, and someone with a working python mapmerge can easily get rid of the temporary types in separate PR.

PsiOmegaDelta commented on code/modules/modular_computers/file_system/programs/camera.dm:
```diff
@@ -0,0 +1,220 @@
+		if(NETWORK_COMMAND || NETWORK_TELECOM)
+			return access_heads
+		if(NETWORK_CRESCENT || NETWORK_ERT)
+			return access_cent_specops
```
What about Torch camera networks?

atlantiscze replied:
Have i missed any specific ones? Most of them should be in the list. Adding more would be trivial.

PsiOmegaDelta replied:
What if the same network on different maps should have different access requirements?
What if someone adds a new network and forgets to modify this file?
What if this is a horrendously hard-coded design?

atlantiscze replied:
Per-map accesses are not supported at this time. To my knowledge, no camera network actually requires it.

If someone adds a new network and forgets to modify this file the network will be using default access_security.

If you have any suggestions on how to better do this, i'll gladly hear them.

PsiOmegaDelta replied:
An associative list of network to permissions, or something of the sort would be more easily manipulated by different maps. However, I'll just add it to my own todo list and do it myself.

PsiOmegaDelta commented:
The Java merger is rumored to now work with all Torch maps.

Raptor1628 commented:
Someone suggested that successfully mapmerging all maps once with the python one would fix things, so who knows? Either way, psyched for this.

PsiOmegaDelta commented:
Can confirm that the Java merger runs perfectly fine now.

atlantiscze commented:
That's excellent news. I'll update this (and the second PR that suffered same problem) once i'm on a normal computer (Wednesday).

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-06-22-FixedCameras

atlantiscze committed:
Removes placeholders
- As java mapmerge works for torch now they were no longer necessary.

atlantiscze commented:
PR updated, no longer uses placeholders as mapmerge works. Ready to merge.

PsiOmegaDelta commented:
Merge conflicts.

atlantiscze committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into 2016-06-22-FixedCameras

atlantiscze commented:
The conflict should be resolved.

PsiOmegaDelta reviewed (approved):
Hmm...
